graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,372 @@
+package com.apollographql.federation.graphqljava.caching;
+
+import com.apollographql.federation.graphqljava._Entity;
+import graphql.ExecutionResult;
+import graphql.ExecutionResultImpl;
+import graphql.execution.instrumentation.InstrumentationContext;
+import graphql.execution.instrumentation.InstrumentationState;
+import graphql.execution.instrumentation.SimpleInstrumentation;
+import graphql.execution.instrumentation.SimpleInstrumentationContext;
+import graphql.execution.instrumentation.parameters.InstrumentationExecutionParameters;
+import graphql.execution.instrumentation.parameters.InstrumentationFieldParameters;
+import graphql.schema.*;
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * A GraphQL Java Instrumentation that computes a max age for an operation based on @cacheControl
+ * directives. See https://www.apollographql.com/docs/apollo-server/performance/caching/
+ */
+public class CacheControlInstrumentation extends SimpleInstrumentation {
+  private final int defaultMaxAge;
+  private final HintCache hintCache = new HintCache();
+
+  private static final String DIRECTIVE_NAME = "cacheControl";
+  private static final String EXTENSION_KEY = "cache-control";
+
+  public CacheControlInstrumentation() {
+    this(0);
+  }
+
+  public CacheControlInstrumentation(int defaultMaxAge) {
+    this.defaultMaxAge = defaultMaxAge;
+  }
+
+  @Override
+  public InstrumentationState createState() {
+    return new CacheControlState();
+  }
+
+  @Override
+  public InstrumentationContext<ExecutionResult> beginField(
+      InstrumentationFieldParameters parameters) {
+    CacheControlState state = parameters.getInstrumentationState();
+    CacheControlPolicy fieldPolicy = new CacheControlPolicy();
+    boolean inheritMaxAge = false;
+
+    GraphQLUnmodifiedType unwrappedReturnType =
+        GraphQLTypeUtil.unwrapAll(parameters.getExecutionStepInfo().getType());
+
+    // There's no way to set a cacheControl directive on the _entities field or the _Entity union in
+    // SDL.
+    // Instead, we can determine the possible concrete types from the representations arguments and
+    // select the
+    // most restrictive cache policy from those types.
```

Comment:
Federation noob here but feels like this is a more general problem about interfaces/unions and this could happen for any interface/union, not just `_Entity`?

```graphql
interface Animal

type Cat implements Animal @cacheControl(maxAge: 60)

type Query {
  animal {
    name
  }
}
```
if the returned animal is a `Cat`, should we use `maxAge: 60`?

Reply:
So, we don't implement this in the TS version. If I recall properly I considered implementing this as part of the semi rewrite in Apollo Server 3 earlier this year (plus the goal of supporting `Query._entities` better) and there wasn't a great way in graphql-js to instrument the "interface type resolution" step appropriately.  Perhaps graphql-java makes this easier, but it would probably be better to stay compatible.  We could come up with a way to get it into graphql-js (perhaps by extending it and requiring a new version of graphql-js for it to work) later perhaps.

Also, because we support dynamic cache control in addition to static cache control, and the dynamic cache control methods can be called from an interface type resolver, there at least is a way to do what you want on an ad hoc per interface level; I hope we are able to get dynamic cache control here too!